lib/bedrock/data_plane/storage/olivine/version_manager/page.ex

```diff
@@ -0,0 +1,290 @@
+
+    encoded_versions = for version <- page.versions, into: <<>>, do: version
+
+    header = <<
+      page.id::unsigned-big-64,
+      page.next_id::unsigned-big-64,
+      key_count::unsigned-big-32,
+      last_key_offset::unsigned-big-32,
+      0::unsigned-big-64
+    >>
+
+    <<header::binary, encoded_versions::binary, encoded_keys::binary>>
+  end
+
+  @spec encode_keys(keys :: [binary()]) :: binary()
+  defp encode_keys(keys), do: for(key <- keys, into: <<>>, do: <<byte_size(key)::unsigned-big-16, key::binary>>)
+
+  @spec from_binary(binary()) :: {:ok, page()} | {:error, :invalid_page}
+  def from_binary(binary) do
+    with <<id::unsigned-big-64, next_id::unsigned-big-64, key_count::unsigned-big-32, _last_key_offset::unsigned-big-32,
+           _reserved::unsigned-big-64, encoded_versions::binary-size(key_count * 8), encoded_keys::binary>> <- binary,
+         {:ok, versions} <- decode_versions(encoded_versions),
+         {:ok, keys} <- decode_keys(encoded_keys) do
+      {:ok,
+       %{
+         id: id,
+         next_id: next_id,
+         keys: keys,
+         versions: versions
+       }}
+    else
+      _ -> {:error, :invalid_page}
+    end
+  end
+
+  @spec decode_keys(encoded_keys :: binary()) :: {:ok, [binary()]} | {:error, :invalid_keys}
+  def decode_keys(encoded_keys), do: decode_keys(encoded_keys, [])
+
+  defp decode_keys(<<key_len::unsigned-big-16, key::binary-size(key_len), rest::binary>>, acc),
+    do: decode_keys(rest, [key | acc])
+
+  defp decode_keys(<<>>, acc), do: {:ok, Enum.reverse(acc)}
+
+  defp decode_keys(_, _acc), do: {:error, :invalid_keys}
+
+  @spec decode_versions(encoded_versions :: binary()) :: {:ok, [Bedrock.version()]} | {:error, :invalid_versions}
+  def decode_versions(encoded_versions), do: decode_versions(encoded_versions, [])
+
+  defp decode_versions(<<version::binary-8, rest::binary>>, acc), do: decode_versions(rest, [version | acc])
+  defp decode_versions(<<>>, acc), do: {:ok, Enum.reverse(acc)}
+  defp decode_versions(_, _acc), do: {:error, :invalid_versions}
+
+  @spec key_count(page :: page()) :: non_neg_integer()
+  def key_count(page), do: length(page.keys)
+
+  @spec keys(page :: page()) :: [binary()]
+  def keys(page), do: page.keys
+
+  @spec lookup_key_in_page(page :: page(), key :: binary()) :: {:ok, Bedrock.version()} | {:error, :not_found}
+  def lookup_key_in_page(page, key) do
+    case Enum.find_index(page.keys, &(&1 == key)) do
+      nil -> {:error, :not_found}
+      index -> {:ok, Enum.at(page.versions, index)}
+    end
+  end
+
+  @spec add_key_to_page(page :: page(), key :: binary(), version :: Bedrock.version()) :: page()
+  def add_key_to_page(page, key, version) when is_binary(version) and byte_size(version) == 8 do
+    case find_insertion_point(page.keys, key) do
+      {index, :duplicate} ->
+        new_versions = List.replace_at(page.versions, index, version)
```

Comment:
How do you find the performance of these List operations? If I change a key 10000 times in the window, I'm gathering that the list of versions will be 10000 long, and I wonder if these start becoming bottlenecks?

Reply:
poor! ;)

the idea is to keep things simple and work out the kinks of the overall idea, and then to dial in the performance once i have a better feel for how it works. i'm just not smart enough to get there in one step. things should improve dramatically when i switch over to using the binary representation and random access is possible.

roll it around, i'm very open to suggestions.

Reply:
i was thinking, though, that one interesting approach might be to do a pass where the mutations are distributed out to the pages that they affect, and then they could be applied all at once, per-page... so better cache locality and fewer (maybe just one) list operation per page. for range-clears, though, you'd do those _immediately_, and any unapplied mutations on pages that get cleared would just get blown away with the page... no need to do _anything_.